examples/perplexity/perplexity.cpp

```diff
@@ -1022,6 +1038,38 @@ static std::vector<winogrande_entry> load_winogrande_from_csv(const std::string&
         auto choice2 = line.substr(comma_pos[2]+1, comma_pos[3] - comma_pos[2] - 1);
         auto answer  = line.substr(comma_pos[3]+1, line.size() - comma_pos[3] - 1);
         auto index = line.substr(0, comma_pos[0]);
+        if ('a' <= sentence[0] && sentence[0] <= 'z') {
+            // make the first letter a capital letter
+            sentence[0] -= 'a' - 'A';
+        }
+        for (int i = 0; i < (int) sentence.size() - 1; ++i) {
+            // trim repeated spaces and spaces before punctuation
+            if (sentence[i] == ' ') {
+                char next = sentence[i+1];
+                if (next == ' ' || next == ',' || next == '.' || next == '\'') {
+                    char r[2] = { next, 0 };
+                    sentence.replace(i, 2, r);
+                    --i; // stay at the same index for repeated spaces
+                }
+            } else if (sentence[i] == ',' || sentence[i] == '.') {
+                if (sentence[i] == sentence[i+1]) {
+                    // trim repeated punctuation (forward to work at the end of sentences)
+                    char r[2] = { sentence[i], 0 };
+                    sentence.replace(i, 2, r);
+                    --i; // same index to then run the other checks on that punctuation
+                } else if (0 < i && sentence[i-1] == sentence[i]) {
+                    // trim repeated punctuation (looks back to work with the space trim)
+                    char r[2] = { sentence[i], 0 };
+                    sentence.replace(i-1, 2, r);
+                    i -= 2; // go back because content was shifted
+                } else if (sentence[i+1] != ' ') {
+                    // add missing space after punctuation
+                    // (since the loop stops before the end, this adds no trailing space)
+                    char r[3] = { sentence[i], ' ', 0 };
+                    sentence.replace(i, 1, r);
+                }
+            }
+        }
```

Comment:
It's somewhat related, because I've changed how the Winogrande score is calculated (it didn't use the correct logits for the choice word of the second choice before, and it used a choice-word-skipping logic to mitigate the effects), and normalizing the sentences helped to make Phi-2_Q4_K_M's score a bit better (`72.6125 +/- 1.2533` when normalizing vs `72.0600 +/- 1.2611` when not normalizing). Some of the sentences in the Winogrande dataset I used has weird punctuation and inconsistent spacing, which sometimes slightly skews the log-likelyhood, but this only affects 111/1267 of the tasks in <https://huggingface.co/datasets/ikawrakow/winogrande-eval-for-llama.cpp>.

```console
$ grep -E "(^[a-z]|\.\w|\.\.| '| \.| ,|  +)" winogrande-debiased-eval.csv | wc -l
111
```

I could try extracting this into another PR or removing it altogether if it's too out-of-place, like if it's better to keep the dataset as-is even if some of the formatting confuses some models.

Reply:
I can review the changes in llama.cpp and ggml, but I cannot review these changes. Since `perplexity` is used to test the correctness of many other things, I think it is important that the changes are carefully reviewed. So as it is, either someone else will have to review this PR, or it will have to be split.

Reply:
In this case I could make Winogrande work as it does on `master`, even though it's wrong and needlessly complicated, and then fix it in a separate PR. All of the benchmarks in `perplexity` (except Winogrande, but I could fix this) should give the exact same results in this PR as on `master`, which *should* make it relatively easy to test that it works as expected or know if I made a mistake somewhere.

Reply:
"Wrong and needlessly complicated": this basically necessitates a separate PR.

Reply:
I have been testing hellaswag since there are some changes, and over 400 tasks with I see a difference of 0.25 compared to master (75.00 vs 74.75) with llama 7B F16. Over mistral F16 there is also a small difference (81.25 vs 81.00). Over the full 10042 tasks the result is basically the same so it is probably nothing, but I wonder if something changed that could result in a different computation? I thought it could be due to different matrix sizes, but there is also a difference with MMQ.

Reply:
> I wonder if something changed that could result in a different computation?

I've tested with `stories260K.gguf` while showing the logprobs of the 4 choices, and I see that the numbers (compared to `master`) are exactly the same for the first choice, while for the other three, the 4 most significant digits match while the rest seem to vary. So if this changes the result, the choices were already very close to each other. This is either due to some numerical instability in floats or there's something I did wrong. (apparently, it's both)

Okay, while typing this I ran some tests, and I don't understand why, but evaluating the last tokens of the sequences (even without extracting the logits) is apparently necessary to get the exact same numbers. Previously, I simply skipped the evaluation of the last tokens because their output isn't necessary at all. It's weird because the choices are in separate sequences, so the KV cache shouldn't be affected... (tested with Mamba too, and at least my intuition holds there, so the way the KV cache is updated for Transformer-based models makes this different. Why would the KV cells of unrelated sequences be changed slightly like this?) I'll need to run tests for the other benchmarks and will push a fix tomorrow.

Reply:
See this comment (and the one referenced in it) for explanation of why the results are different: https://github.com/ggerganov/whisper.cpp/issues/1941#issuecomment-1986923227